Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,81 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         }
     }
 
+    /// Replaces the specified subrange of elements with the given collection.
+    ///
+    /// - Parameter subrange:
+    /// The subrange of the collection to replace. The bounds of the range must be valid indices of the collection.
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: Range<Index>, with newElements: C) where C : Collection, E == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        let lowerBound = bufferIndex(ofIndex: subrange.lowerBound)
+        let upperBound = bufferIndex(ofIndex: subrange.upperBound == self.endIndex ? subrange.upperBound - 1 : subrange.upperBound)
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(stride(from: subrange.lowerBound, to: subrange.upperBound, by: 1), newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            #if swift(>=4.2)
+            let capacityDelta = (newElements.count - subrange.count)
+            #else
+            let capacityDelta = (Int(newElements.count) - subrange.count)
+            #endif
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(newCapacity)
+            var index = self.headIdx
+            while index != self.tailIdx {
+                if index == lowerBound {
+                    for element in newElements {
+                        newBuffer.append(element)
+                    }
+                    while index < upperBound {
+                        index = self.bufferIndex(after: index)
+                    }
+                } else {
+                    newBuffer.append(self.buffer[index])
+                    index = self.bufferIndex(after: index)
+                }
+            }
+
+            self.tailIdx = newBuffer.count
+            let repetitionCount = self.buffer.count - newBuffer.count
+            if repetitionCount > 0 {
+                newBuffer.append(contentsOf: repeatElement(nil, count: repetitionCount))
+            }
+            self.headIdx = 0
+            self.buffer = newBuffer
+        }
+    }
+
+    public mutating func removeSubrange(_ bounds: Range<Int>) {
```

Comment:
This should probably precondition that the `bounds` are in fact valid indices for the collection.